base/strings/basic.jl

```diff
@@ -321,7 +321,14 @@ julia> "abc" == "αβγ"
 false
 ```
 """
-==(a::AbstractString, b::AbstractString) = cmp(a, b) == 0
+function ==(a::AbstractString, b::AbstractString)
+    a === b && return true
+    for (c::AbstractChar, d::AbstractChar) in zip(a, b)
+        c != d && return false
+    end
+    length(a) != length(b) && return false
```

Comment:
These are quite likely O(n) again. We should be able to get away with only one O(N) iteration.

Reply:
@omus, maybe change back to the iterator based approach to fix this?